### PR TITLE
Implement @impl directive

### PR DESCRIPTION
**Pull request by samphilipd**
Implement @impl directive

    - See: #5734

### CODE REVIEW

File: lib/elixir/lib/kernel.ex [hunk 1]
```diff
@@ -2476,7 +2476,7 @@ defmodule Kernel do
         quote do: Module.put_attribute(__MODULE__, unquote(name), unquote(arg),
                                        unquote(stack), unquote(line))
 
-      :lists.member(name, [:moduledoc, :typedoc, :doc]) ->
+      :lists.member(name, [:moduledoc, :typedoc, :doc, :impl]) ->
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: It is better to not rely on the line number. That's not a general feature of Elixir module attributes and as such it should be relied on very rarely.

**Comment by josevalim** [comment]
According to the build it is working: https://travis-ci.org/elixir-lang/elixir/jobs/224857673#L237

It is just failing elsewhere inside the `check_impl` function.

**Comment by samphilipd** [comment]
@josevalim This is about as far as I can go on my own before getting feedback, would you mind taking a look?

File: lib/elixir/lib/module.ex [hunk 2]
```diff
@@ -1097,14 +1106,120 @@ defmodule Module do
...
+      :ok
+    end
+  end
+  defp do_check_impl(_module, _kind, _function_tuple, value, _) when value == true do
+    raise "Implement me"
+  end
+  defp do_check_impl(_module, _kind, function_tuple, behaviour, _) when is_atom(behaviour) do
+    cond do
+      not Code.ensure_compiled?(behaviour) ->
+        {:error, "got @impl #{inspect(behaviour)} but #{inspect(behaviour)} " <>
```
Comment (samphilipd) [inline lib/elixir/lib/module.ex]: Perhaps we should actually raise instead of warn here?

File: lib/elixir/lib/module.ex [hunk 3]
```diff
@@ -1158,6 +1273,18 @@ defmodule Module do
         :ok
     end
 
+    case {key, value} do
+      {:impl, nil} -> :ok
+      {:impl, false} -> :ok
+      {:impl, true} -> raise "TODO: implement"
```
Comment (samphilipd) [inline lib/elixir/lib/module.ex]: @josevalim The original proposal specified an `@impl true` variant but I actually prefer the explicitness of `@impl Module`.

Perhaps we could ship the specific case and see how people like it. If it seems obvious that `@impl true` would be nice to have, we can easily add it later. Adding it now and removing it later would be much more difficult.

Comment (josevalim) [inline lib/elixir/lib/module.ex]: @samphilipd I am already decided on making `@impl true` part of the implementation so we should go ahead with it.

File: lib/elixir/lib/module.ex [hunk 4]
```diff
@@ -1097,14 +1106,120 @@ defmodule Module do
...
+  # Used internally to check impl usage.
+  # This function is private and must be used only internally.
+  def check_impl(env, kind, name, args, _guards, _body) do
+    module = env.module
+    table = data_table_for(module)
+
+    value = get_impl_info(table)
+
+    line = env.line
+    file = env.file
```
Comment (josevalim) [inline lib/elixir/lib/module.ex]: Please use pattern matching to extract line, file, module and others out.

    %{module: module, line: line, file: file} = env

File: lib/elixir/lib/module.ex [hunk 5]
```diff
@@ -1158,6 +1273,18 @@ defmodule Module do
         :ok
     end
 
+    case {key, value} do
+      {:impl, nil} -> :ok
+      {:impl, false} -> :ok
```
Comment (josevalim) [inline lib/elixir/lib/module.ex]: Why would we need to change `store_typespec`? :)

File: lib/elixir/lib/module.ex [hunk 6]
```diff
@@ -1174,6 +1301,42 @@ defmodule Module do
...
+    end
+  end
+
+  defp get_behaviours(module) do
+    get_attribute(module, :behaviour)
+  end
+
+  # TODO: address "module attribute @impl_required was set but never used" warnings
+  defp set_impl_required(table, unread_line) do
+    case :ets.lookup(table, :impl_required) do
```
Comment (josevalim) [inline lib/elixir/lib/module.ex]: For internal attributes, such as `:impl_required`, you can namespace it as`{:elixir, :impl_required}` to avoid conflicts with user attributes.

File: lib/elixir/src/elixir_module.erl [hunk 7]
```diff
@@ -35,6 +35,11 @@ delete_doc(#{module := Module}, _, _, _, _, _) ->
   ets:delete(data_table(Module), doc),
   ok.
 
+delete_impl(#{module := Module}, _, _, _, _, _) ->
+  ets:delete(data_table(Module), impl),
+  % TODO: What to do here?
```
Comment (josevalim) [inline lib/elixir/src/elixir_module.erl]: Nothing. :) This is during bootstrap, so we don't care. If there are warnings, the modules will be recompiled after bootstrap, so we get them anyway.

File: README.md [hunk 8]
```diff
@@ -18,6 +18,24 @@ cd elixir
...
+```sh
+make clean_exbeam compile
+```
+
+Or to rebuild everything without running tests:
+
+```sh
+make clean compile
+```
+
```
Comment (josevalim) [inline README.md]: Can you please send those changes in another PR? :)

File: lib/elixir/lib/module.ex (hunk 5 above)
Comment (samphilipd) [inline lib/elixir/lib/module.ex]: You are correct. It's not necessary.

**Comment by samphilipd** [comment]
@josevalim I addressed your comments, this is ready for a second round of feedback.

File: lib/elixir/lib/module.ex [hunk 9]
```diff
@@ -1097,14 +1138,197 @@ defmodule Module do
...
+    Enum.join(options, ", ") <> last_words_connector <> last
+  end
+
+  defp msg_impl_true_but_function_is_not_callback(function_tuple = {name, _arity}, callbacks_with_behaviour, behaviours) do
+    callbacks_with_same_name =
+      for (tuple = {callback_name, _callback_arity}) <- :maps.keys(callbacks_with_behaviour),
+      callback_name == name,
+      do: tuple
+
+    case length(callbacks_with_same_name) do
```
Comment (samphilipd) [inline lib/elixir/lib/module.ex]: There's probably room for refactoring here to make it neater.

File: lib/elixir/test/elixir/kernel/impl_test.exs [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+      @behaviour Kernel.ImplTest.Behaviour
+
+      @impl Kernel.ImplTest.Behaviour
+      def foo() do
+        :ok
+      end
+    end
+
+    # TODO: This always returns nil... how to test it?
+    assert Code.get_docs(Kernel.ImplTest.ImplAttributes, :docs) == false
```
Comment (samphilipd) [inline lib/elixir/test/elixir/kernel/impl_test.exs]: This test doesn't work. Is there another way to get the `@doc` of a module?

File: lib/elixir/lib/module.ex [hunk 11]
```diff
@@ -1097,14 +1138,197 @@ defmodule Module do
...
+      set_impl_required(table, line)
+      set_doc_false_if_unset(table, line)
+    end
+
+    arity = length(args)
+    pair = {name, arity}
+    required? = impl_required_for_callbacks?(table)
+
+    case do_check_impl(module, kind, pair, value, required?) do
+      {:error, message} -> :elixir_errors.warn line, file, message
```
Comment (samphilipd) [inline lib/elixir/lib/module.ex]: This should always be a warning rather than an outright ArgumentError?

File: lib/elixir/lib/module.ex [hunk 12]
```diff
@@ -1097,14 +1138,197 @@ defmodule Module do
...
+  end
+  defp do_check_impl(_, kind, _, _, _) when kind in [:type, :typep] do
+    {:error, "@impl is always ignored for type annotations"}
+  end
+  defp do_check_impl(_, _, _, value, impl_required_for_callbacks?)
+       when value in [nil, false] and not impl_required_for_callbacks? do
+    :ok
+  end
+  defp do_check_impl(module, _, function_tuple, value, impl_required_for_callbacks?)
+       when value in [nil, false] and impl_required_for_callbacks? do
```
Comment (samphilipd) [inline lib/elixir/lib/module.ex]: It might make sense to merge this somehow with the `true` and behaviour cases since they all duplicate this `callbacks_with_behaviour` logic.

File: lib/elixir/lib/module.ex [hunk 13]
```diff
@@ -1097,14 +1138,197 @@ defmodule Module do
...
+        else
+          {
+            :error,
+            msg_impl_behaviour_but_function_is_not_callback(function_tuple, callbacks, behaviour),
+          }
+        end
+    end
+  end
+
+  defp to_sentence(options, last_words_connector \\ " or ")
```
Comment (samphilipd) [inline lib/elixir/lib/module.ex]: This `to_sentence` function seems a bit too much for `Module`, but I couldn't think of another way to nicely present the warnings.

**Comment by samphilipd** [comment]
@josevalim OK I'm fairly happy with this, only two questions:

1. Should we ever raise in any case? Or always warn?
2. See the test at the bottom of warnings_test.exs, I can't see how to implement this case.

**Comment by samphilipd** [comment]
@josevalim just a polite nudge for a review on this :)

File: lib/elixir/lib/module.ex [hunk 14]
```diff
@@ -53,6 +53,40 @@ defmodule Module do
...
+        @impl true
+        def default_port(), do: 80
+
+        @impl true
+        def parse(info), do: info
+      end
+
+  You may pass either `false`, `true`, or a specific behaviour to `@impl`.
+
+    defmodule Foo do
```
Comment (josevalim) [inline lib/elixir/lib/module.ex]: Please indent with four spaces.

**Comment by josevalim** [comment]
Thank you @samphilipd! I will merge it locally and tidy up what is left. :)

**Comment by josevalim** [comment]
This has been merged and push to master. In order for the last test to pass, we had to change when the check happens, and this one done in a another commit: 92e5d041133241ca7309466bdfda6b72dc5b55f7

Thank you so much!  :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
